src/phase-2/scWETHv2.sol

```diff
@@ -0,0 +1,533 @@
+        return protocolAdapters.contains(_adapterId);
+    }
+
+    function getAdapter(uint256 _adapterId) external view returns (address adapter) {
+        (, adapter) = protocolAdapters.tryGet(_adapterId);
+    }
+
+    /// @notice returns the total assets (WETH) held by the strategy
+    function totalAssets() public view override returns (uint256 assets) {
+        // value of the supplied collateral in eth terms using chainlink oracle
+        assets = totalCollateral();
+
+        // subtract the debt
+        assets -= totalDebt();
+
+        // add float
+        assets += asset.balanceOf(address(this));
+    }
+
+    /// @notice returns the total assets supplied as collateral (in WETH terms)
+    function totalCollateral() public view returns (uint256 collateral) {
+        uint256 n = protocolAdapters.length();
+        address adapter;
+        for (uint256 i; i < n; i++) {
+            (, adapter) = protocolAdapters.at(i);
+            collateral += IAdapter(adapter).getCollateral(address(this));
+        }
+
+        collateral = oracleLib.wstEthToEth(collateral);
+    }
+
+    /// @notice returns the total ETH borrowed
+    function totalDebt() public view returns (uint256 debt) {
+        uint256 n = protocolAdapters.length();
+        address adapter;
+        for (uint256 i; i < n; i++) {
+            (, adapter) = protocolAdapters.at(i);
+            debt += IAdapter(adapter).getDebt(address(this));
+        }
+    }
+
+    //////////////////// EXTERNAL METHODS //////////////////////////
+
+    /// @notice helper method to directly deposit ETH instead of weth
+    function deposit(address receiver) external payable returns (uint256 shares) {
+        uint256 assets = msg.value;
+
+        // Check for rounding error since we round down in previewDeposit.
+        require((shares = previewDeposit(assets)) != 0, "ZERO_SHARES");
+
+        // wrap eth
+        WETH(payable(address(asset))).deposit{value: assets}();
+
+        _mint(receiver, shares);
+
+        emit Deposit(msg.sender, receiver, assets, shares);
+
+        afterDeposit(assets, shares);
+    }
+
+    function withdraw(uint256, address, address) public virtual override returns (uint256) {
+        revert PleaseUseRedeemMethod();
+    }
+
+    function redeem(uint256 shares, address receiver, address owner) public override returns (uint256 assets) {
+        if (msg.sender != owner) {
+            uint256 allowed = allowance[owner][msg.sender]; // Saves gas for limited approvals.
+
+            if (allowed != type(uint256).max) allowance[owner][msg.sender] = allowed - shares;
+        }
+
+        // Check for rounding error since we round down in previewRedeem.
+        require((assets = previewRedeem(shares)) != 0, "ZERO_ASSETS");
+
+        beforeWithdraw(assets, shares);
+
+        _burn(owner, shares);
+
+        uint256 balance = asset.balanceOf(address(this));
+
+        if (assets > balance) {
```

Comment:
What could cause this condition become `true`?

Reply:
During withdrawing everything, the actual withdrawn amount might be less than totalAsssets (due to slippage incurred during wstEth to weth swap)